models/battleship/game_manager.go

```diff
@@ -0,0 +1,162 @@
+package battleship
+
+import (
+	"github.com/google/uuid"
+	cerr "github.com/saeidalz13/battleship-backend/internal/error"
+
+	"sync"
+)
+
+type GameManager interface {
+	CreateGame(difficulty uint8) (*Game, error)
+	GetGame(gameUuid string) (*Game, error)
+	GetPlayer(game *Game, isHost bool) *BattleshipPlayer
+	GetOtherPlayerForGame(game *Game, player Player) *BattleshipPlayer
+	GetGameUuid(game *Game) string
+	GetGameDifficulty(game *Game) uint8
+	CreateHostPlayerForGame(game *Game, sessionId string) *BattleshipPlayer
+	CreateJoinPlayerForGame(game *Game, sessionId string) *BattleshipPlayer
+	SetPlayerReadyForGame(game *Game, player Player, selectedGrid Grid) error
+	AreAttackCoordinatesValid(game *Game, coordinates Coordinates) bool
+	CallRematchForGame(game *Game)
+	ResetRematchForGame(game *Game) error
+	IsRematchAlreadyCalled(game *Game) bool
+	IsGameReadyToStart(game *Game) bool
+	GetJoinPlayerSunkenShips(game *Game) uint8
+	GetHostPlayerSunkenShips(game *Game) uint8
+
+	isDifficultyValid(uint8) bool
+}
+
+type BattleshipGameManager struct {
+	games map[string]*Game
+	mu    sync.RWMutex
+}
+
+var _ GameManager = (*BattleshipGameManager)(nil)
+
+func NewBattleshipGameManager() *BattleshipGameManager {
+	return &BattleshipGameManager{
+		games: make(map[string]*Game, 10),
+	}
+}
+func (bgm *BattleshipGameManager) CreateGame(difficulty uint8) (*Game, error) {
+	if !bgm.isDifficultyValid(difficulty) {
+		return nil, cerr.ErrInvalidGameDifficulty()
+	}
+
+	gameUuid := uuid.NewString()[:6]
+	bgm.games[gameUuid] = newGame(difficulty, gameUuid)
+
+	return bgm.games[gameUuid], nil
+}
+
+func (bgm *BattleshipGameManager) GetGame(gameUuid string) (*Game, error) {
+	bgm.mu.RLock()
+	game, prs := bgm.games[gameUuid]
+	bgm.mu.RUnlock()
+	if !prs {
+		return nil, cerr.ErrGameNotExists(gameUuid)
+	}
+
+	if game == nil {
+		bgm.mu.Lock()
+		delete(bgm.games, gameUuid)
+		bgm.mu.Unlock()
+		return nil, cerr.ErrGameIsNil(gameUuid)
+	}
+
+	return game, nil
+}
+
+func (bgm *BattleshipGameManager) GetPlayer(game *Game, isHost bool) *BattleshipPlayer {
+	if isHost {
+		return game.hostPlayer
+	}
+
+	return game.joinPlayer
+}
+
+func (bgm *BattleshipGameManager) GetOtherPlayerForGame(game *Game, player Player) *BattleshipPlayer {
```

Comment:
We could pass `isHost` here too just like `GetPlayer` function instead of passing the whole `Player` object